### PR TITLE
Update CODEOWNERS

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

- Restructure sections, from most general to most specific
- Add myself to some sections
- Add @Borda to some sections
- Add all packages. Assigned members to these new sections.
- Add @tchaton to API so non-API changes in these files are easier to approve. For example, docstring updates or minor bugs.

cc @Borda

### CODE REVIEW

**Review by Borda — APPROVED** [review]

**Comment by tchaton** [comment]
Hey @carmocca,

Would you mind adding Justus to `/pytorch_lightning/core` and both `@awaelchli @justusschock` for Trainer. 

Best regards,
T.C

**Review by tchaton — APPROVED** [review]

File: .github/CODEOWNERS [hunk 1]
```diff
@@ -5,48 +5,48 @@
...
+/pytorch_lightning/metrics/             @teddykoker @ananyahjha93 @justusschock
+/tests/metrics/                         @teddykoker @ananyahjha93 @justusschock
+/docs/source/metrics.rst                @teddykoker @ananyahjha93 @justusschock
+
+# API
+/pytorch_lightning/callbacks/base.py  @williamfalcon @tchaton
+/pytorch_lightning/core/datamodule.py @williamfalcon @tchaton
+/pytorch_lightning/trainer/trainer.py @williamfalcon @tchaton
+/pytorch_lightning/core/hooks.py      @williamfalcon @tchaton
+/pytorch_lightning/core/lightning.py  @williamfalcon @tchaton
```
Comment (rohitgr7) [inline .github/CODEOWNERS:49]: can you guys add more (maybe 1 more) here??

Comment (carmocca) [reply .github/CODEOWNERS:49]: I'm not sure if @williamFalcon would be okay with that

Comment (justusschock) [reply .github/CODEOWNERS:49]: I think, that's exactly what @tchaton requested above?

**Review by justusschock — APPROVED** [review]

File: .github/CODEOWNERS (hunk 1 above)
Comment (awaelchli) [reply .github/CODEOWNERS:49]: if only william and thomas are listed, will we still get request for review?

File: .github/CODEOWNERS [hunk 2]
```diff
@@ -5,48 +5,48 @@
...
+/pytorch_lightning/distributed          @williamfalcon @tchaton @awaelchli
+/pytorch_lightning/loggers              @tchaton @awaelchli @borda
+/pytorch_lightning/overrides            @tchaton @SeanNaren @borda
+/pytorch_lightning/plugins              @tchaton @SeanNaren @awaelchli @justusschock
+/pytorch_lightning/profiler             @williamfalcon @tchaton @borda
+/pytorch_lightning/trainer              @williamfalcon @borda @tchaton @SeanNaren @carmocca @awaelchli @justusschock
+/pytorch_lightning/trainer/connectors   @tchaton @SeanNaren @carmocca @borda
+/pytorch_lightning/tuner                @SkafteNicki @borda @awaelchli
+/pytorch_lightning/utilities            @borda @tchaton @SeanNaren @carmocca
+# Metrics
```
Comment (awaelchli) [inline .github/CODEOWNERS:39]: the other sections have a blank line above

Comment (Borda) [reply .github/CODEOWNERS:39]: nop

File: .github/CODEOWNERS (hunk 1 above)
Comment (carmocca) [reply .github/CODEOWNERS:50]: Not for these files

Comment (awaelchli) [reply .github/CODEOWNERS:50]: that's unfortunate :(

Comment (tchaton) [reply .github/CODEOWNERS:50]: Let's add @SeanNaren there and see if Will accepts this update.

**Review by tchaton — APPROVED** [review]
LGTM !

File: .github/CODEOWNERS (hunk 1 above)
Comment (Borda) [reply .github/CODEOWNERS:50]: @awaelchli we can add you and me as a reviewer by merfigy even we won't be code-owners... :]

File: .github/CODEOWNERS [hunk 3]
```diff
@@ -5,48 +5,49 @@
...
+/pytorch_lightning/metrics/             @teddykoker @ananyahjha93 @justusschock
+/tests/metrics/                         @teddykoker @ananyahjha93 @justusschock
+/docs/source/metrics.rst                @teddykoker @ananyahjha93 @justusschock
+
+# API
+/pytorch_lightning/callbacks/base.py  @williamfalcon @tchaton
+/pytorch_lightning/core/datamodule.py @williamfalcon @tchaton
+/pytorch_lightning/trainer/trainer.py @williamfalcon @tchaton
+/pytorch_lightning/core/hooks.py      @williamfalcon @tchaton
+/pytorch_lightning/core/lightning.py  @williamfalcon @tchaton
```
Comment (Borda) [inline .github/CODEOWNERS:50]:
```suggestion
/pytorch_lightning/callbacks/base.py  @williamfalcon @tchaton @SeanNaren @Borda
/pytorch_lightning/core/datamodule.py @williamfalcon @tchaton @SeanNaren @Borda
/pytorch_lightning/trainer/trainer.py @williamfalcon @tchaton @SeanNaren @Borda
/pytorch_lightning/core/hooks.py      @williamfalcon @tchaton @SeanNaren @Borda
/pytorch_lightning/core/lightning.py  @williamfalcon @tchaton @SeanNaren @Borda
```
cc: @tchaton

File: .github/CODEOWNERS (hunk 1 above)
Comment (Borda) [reply .github/CODEOWNERS:50]: @awaelchli have look at https://docs.mergify.io/conditions.html#how-to-match-lists
```
files=README is true if the file README is modified in the pull request.
files!=README is true if the file README is not modified in the pull request.
files~=^src/ is true if any files in the src directory is modified in the pull request.
-files~=^src/ is true if none of the files that are modified are in the src directory.
```

**Comment by edenlightning** [comment]
@carmocca i think this looks good accept for the API files- we still need will to approve them. Is it possible to have both tchaton and williamFalcon approve the API files?

**Comment by carmocca** [comment]
> Is it possible to have both tchaton and williamFalcon approve the API files?

The only reason for adding @tchaton to these API files is to avoid blocking small changes to these files (docs updates, formatting changes, minor bug fixes, non-api refactors, typing improvements...)

This requires putting trust on him (great power, great responsability :spider:). As he would have the power to circumvent @williamFalcon's approval

If you/he are not okay with this. It is better to keep only @williamFalcon as the only CODEOWNER of these files because we wouldn't be gaining anything by requiring both.

**Comment by Borda** [comment]
yeas the main reason is just not to be blocked with simple typos... @edenlightning do you have any other suggestion? 🐰

File: .github/CODEOWNERS [hunk 4]
```diff
@@ -5,48 +5,49 @@
...
+/pytorch_lightning/tuner                @SkafteNicki @borda @awaelchli
+/pytorch_lightning/utilities            @borda @tchaton @SeanNaren @carmocca
+
+# Metrics
+/pytorch_lightning/metrics/             @teddykoker @ananyahjha93 @justusschock
+/tests/metrics/                         @teddykoker @ananyahjha93 @justusschock
+/docs/source/metrics.rst                @teddykoker @ananyahjha93 @justusschock
+
+# API
+/pytorch_lightning/callbacks/base.py  @williamfalcon @tchaton
```
Comment (edenlightning) [inline .github/CODEOWNERS:46]:
```suggestion
/pytorch_lightning/callbacks/base.py  @williamfalcon
```

File: .github/CODEOWNERS [hunk 5]
```diff
@@ -5,48 +5,49 @@
...
+/pytorch_lightning/core/datamodule.py @williamfalcon @tchaton
```
Comment (edenlightning) [inline .github/CODEOWNERS:47]:
```suggestion
/pytorch_lightning/core/datamodule.py @williamFalcon
```

File: .github/CODEOWNERS [hunk 6]
```diff
@@ -5,48 +5,49 @@
...
+/pytorch_lightning/trainer/trainer.py @williamfalcon @tchaton
+/pytorch_lightning/core/hooks.py      @williamfalcon @tchaton
```
Comment (edenlightning) [inline .github/CODEOWNERS:49]:
```suggestion
/pytorch_lightning/core/hooks.py      @williamfalcon
```

**Review by williamFalcon — APPROVED** [review]

**Pull request merged**